src/box/memtx_space.c

```diff
@@ -443,6 +446,69 @@ memtx_space_replace_tuple(struct space *space, struct txn_stmt *stmt,
 	return rc;
 }
 
+/**
+ * Add tuple to space during recovery in MEMTX_OK state, when all indexes
+ * are built.
+ */
+static int
+memtx_space_recovery_insert(struct space *space, struct tuple *tuple)
+{
+	struct index *pk = index_find(space, 0);
+	assert(pk != NULL);
+	struct tuple *successor;
+	struct tuple *old_tuple;
+	if (index_replace(pk, /*old_tuple=*/NULL, tuple, DUP_INSERT,
+			  &old_tuple, &successor) != 0)
+		return -1;
+	assert(old_tuple == NULL);
+	for (uint32_t i = 1; i < space->index_count; i++) {
+		struct index *index = space->index[i];
+		if (index_replace(index, /*old_tuple=*/NULL, tuple,
+				  DUP_INSERT, &old_tuple, &successor) != 0)
+			return -1;
+		assert(old_tuple == NULL);
+	}
+	assert(old_tuple == NULL);
+	memtx_space_update_tuple_stat(space, /*old_tuple=*/NULL, tuple);
+	tuple_ref(tuple);
+	return 0;
+}
+
+int
+memtx_space_recovery_execute_insert(struct space *space,
+				    struct request *request)
+{
+	struct tuple *tuple =
+		space->format->vtab.tuple_new(space->format, request->tuple,
+					      request->tuple_end);
+	if (tuple == NULL)
+		return -1;
+	if (space->format->is_compressed) {
+		struct tuple *compressed_tuple = memtx_tuple_compress(tuple);
+		if (compressed_tuple == NULL)
+			goto error;
+		tuple_delete(tuple);
+		tuple = compressed_tuple;
+	}
+	struct memtx_engine *memtx = (struct memtx_engine *)space->engine;
+	switch (memtx->state) {
+	case MEMTX_INITIAL_RECOVERY:
+		if (memtx_space_build_next(space, tuple) != 0)
+			goto error;
+		break;
+	case MEMTX_OK:
+		if (memtx_space_recovery_insert(space, tuple) != 0)
```

Comment:
Do we really need to handle this case here? Maybe process the 'force recovery' mode through the transaction manager? Function names are really ugly BTW.